### ISSUE TITLE
Plugin throws Non-idempotent computation exception OwO

### ISSUE THREAD

**Issue by ice1000** [intellij-rust/intellij-rust#4495]
## Environment

* **IntelliJ Rust plugin version:** 0.2.107.2754-193-nightly
* **Rust toolchain version:** 1.38.0 (625451e37 2019-09-23) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2019.3 EAP (CL-193.4099.17)
* **Operating system:** Linux 4.15.0-65-generic

## Problem description

```
java.lang.Throwable: Non-idempotent computation: it returns different results when invoked multiple times or on different threads:
  74 != 80
  which is length of [SimpleScopeEntry(name=builtin, element=FILE, subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=HasAttrs, element=RsTraitItemImpl(TRAIT_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), LazyScopeEntry(Attribute, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(ast, FILE), LazyScopeEntry(AttrId, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(AttrStyle, RsEnumItemImpl(ENUM_ITEM)), LazyScopeEntry(Name, null), LazyScopeEntry(Ident, null), LazyScopeEntry(Path, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(PathSegment, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(MetaItem, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(MetaItemKind, RsEnumItemImpl(ENUM_ITEM)), LazyScopeEntry(NestedMetaItem, RsEnumItemImpl(ENUM_ITEM)), LazyScopeEntry(Lit, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(LitKind, RsEnumItemImpl(ENUM_ITEM)), LazyScopeEntry(Expr, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(Item, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(Local, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(Stmt, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(StmtKind, RsEnumItemImpl(ENUM_ITEM)), LazyScopeEntry(GenericParam, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(visit_clobber, null), LazyScopeEntry(BytePos, null), LazyScopeEntry(Spanned, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(dummy_spanned, null), LazyScopeEntry(doc_comment_style, null), LazyScopeEntry(strip_doc_comment_decoration, null), LazyScopeEntry(Parser, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(parse, FILE), LazyScopeEntry(ParseSess, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(PResult, RsTypeAliasImpl(TYPE_ALIAS)), LazyScopeEntry(token, FILE), LazyScopeEntry(Token, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(P, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(sym, null), LazyScopeEntry(Symbol, null), LazyScopeEntry(ThinVec, null), LazyScopeEntry(TokenStream, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(TokenTree, RsEnumItemImpl(ENUM_ITEM)), LazyScopeEntry(DelimSpan, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(GLOBALS, null), LazyScopeEntry(debug, null), LazyScopeEntry(FileName, null), LazyScopeEntry(Span, null), LazyScopeEntry(iter, FILE), LazyScopeEntry(DerefMut, RsTraitItemImpl(TRAIT_ITEM)), STAR_IMPORTS, SimpleScopeEntry(name=proc_macro, element=FILE, subst=org.rust.lang.core.types.EmptySubstitution@0), LazyScopeEntry(std, FILE), LazyScopeEntry(core, FILE), LazyScopeEntry(alloc, FILE), LazyScopeEntry(panic_unwind, FILE), LazyScopeEntry(test, FILE), LazyScopeEntry(panic_abort, FILE), LazyScopeEntry(term, FILE), LazyScopeEntry(unwind, FILE), LazyScopeEntry(build_helper, FILE), LazyScopeEntry(rustc_asan, FILE), LazyScopeEntry(rustc_lsan, FILE), LazyScopeEntry(rustc_msan, FILE), LazyScopeEntry(rustc_tsan, FILE), SimpleScopeEntry(name=AttrError, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=AttributeTemplate, element=RsStructItemImpl(STRUCT_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=InlineAttr, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=OptimizeAttr, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=UnwindAttr, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=Stability, element=RsStructItemImpl(STRUCT_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=StabilityLevel, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=RustcDeprecation, element=RsStructItemImpl(STRUCT_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=Deprecation, element=RsStructItemImpl(STRUCT_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=ReprAttr, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=IntType, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=TransparencyError, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), STAR_IMPORTS] and [SimpleScopeEntry(name=builtin, element=FILE, subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=HasAttrs, element=RsTraitItemImpl(TRAIT_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), LazyScopeEntry(Attribute, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(ast, FILE), LazyScopeEntry(AttrId, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(AttrStyle, RsEnumItemImpl(ENUM_ITEM)), LazyScopeEntry(Name, null), LazyScopeEntry(Ident, null), LazyScopeEntry(Path, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(PathSegment, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(MetaItem, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(MetaItemKind, RsEnumItemImpl(ENUM_ITEM)), LazyScopeEntry(NestedMetaItem, RsEnumItemImpl(ENUM_ITEM)), LazyScopeEntry(Lit, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(LitKind, RsEnumItemImpl(ENUM_ITEM)), LazyScopeEntry(Expr, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(Item, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(Local, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(Stmt, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(StmtKind, RsEnumItemImpl(ENUM_ITEM)), LazyScopeEntry(GenericParam, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(visit_clobber, null), LazyScopeEntry(BytePos, null), LazyScopeEntry(Spanned, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(dummy_spanned, null), LazyScopeEntry(doc_comment_style, null), LazyScopeEntry(strip_doc_comment_decoration, null), LazyScopeEntry(Parser, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(parse, FILE), LazyScopeEntry(ParseSess, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(PResult, RsTypeAliasImpl(TYPE_ALIAS)), LazyScopeEntry(token, FILE), LazyScopeEntry(Token, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(P, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(sym, null), LazyScopeEntry(Symbol, null), LazyScopeEntry(ThinVec, null), LazyScopeEntry(TokenStream, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(TokenTree, RsEnumItemImpl(ENUM_ITEM)), LazyScopeEntry(DelimSpan, RsStructItemImpl(STRUCT_ITEM)), LazyScopeEntry(GLOBALS, null), LazyScopeEntry(debug, null), LazyScopeEntry(FileName, null), LazyScopeEntry(Span, null), LazyScopeEntry(iter, FILE), LazyScopeEntry(DerefMut, RsTraitItemImpl(TRAIT_ITEM)), STAR_IMPORTS, SimpleScopeEntry(name=proc_macro, element=FILE, subst=org.rust.lang.core.types.EmptySubstitution@0), LazyScopeEntry(std, FILE), LazyScopeEntry(core, FILE), LazyScopeEntry(alloc, FILE), LazyScopeEntry(panic_unwind, FILE), LazyScopeEntry(test, FILE), LazyScopeEntry(panic_abort, FILE), LazyScopeEntry(term, FILE), LazyScopeEntry(unwind, FILE), LazyScopeEntry(build_helper, FILE), LazyScopeEntry(rustc_asan, FILE), LazyScopeEntry(rustc_lsan, FILE), LazyScopeEntry(rustc_msan, FILE), LazyScopeEntry(rustc_tsan, FILE), SimpleScopeEntry(name=AttrError, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=AttributeTemplate, element=RsStructItemImpl(STRUCT_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=InlineAttr, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=OptimizeAttr, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=UnwindAttr, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=Stability, element=RsStructItemImpl(STRUCT_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=StabilityLevel, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=RustcDeprecation, element=RsStructItemImpl(STRUCT_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=Deprecation, element=RsStructItemImpl(STRUCT_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=ReprAttr, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=IntType, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=TransparencyError, element=RsEnumItemImpl(ENUM_ITEM), subst=org.rust.lang.core.types.EmptySubstitution@0), STAR_IMPORTS, SimpleScopeEntry(name=ReprInt, element=RsEnumVariantImpl(ENUM_VARIANT), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=ReprC, element=RsEnumVariantImpl(ENUM_VARIANT), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=ReprPacked, element=RsEnumVariantImpl(ENUM_VARIANT), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=ReprSimd, element=RsEnumVariantImpl(ENUM_VARIANT), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=ReprTransparent, element=RsEnumVariantImpl(ENUM_VARIANT), subst=org.rust.lang.core.types.EmptySubstitution@0), SimpleScopeEntry(name=ReprAlign, element=RsEnumVariantImpl(ENUM_VARIANT), subst=org.rust.lang.core.types.EmptySubstitution@0)]
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:145)
	at com.intellij.util.IdempotenceChecker.checkEquivalence(IdempotenceChecker.java:70)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:250)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:43)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:64)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:147)
	at org.rust.lang.core.resolve.ItemResolutionKt.processItemDeclarationsWithCache(ItemResolution.kt:277)
	at org.rust.lang.core.resolve.NameResolutionKt.processLexicalDeclarations(NameResolution.kt:1225)
	at org.rust.lang.core.resolve.NameResolutionKt.access$processLexicalDeclarations(NameResolution.kt:1)
	at org.rust.lang.core.resolve.NameResolutionKt$processNestedScopesUpwards$stop$2.invoke(NameResolution.kt:1354)
	at org.rust.lang.core.resolve.NameResolutionKt$processNestedScopesUpwards$stop$2.invoke(NameResolution.kt)
	at org.rust.lang.core.resolve.NameResolutionKt.walkUp(NameResolution.kt:1463)
	at org.rust.lang.core.resolve.NameResolutionKt.walkUp$default(NameResolution.kt:1456)
	at org.rust.lang.core.resolve.NameResolutionKt.processNestedScopesUpwards(NameResolution.kt:1347)
	at org.rust.lang.core.resolve.NameResolutionKt.processUnqualifiedPathResolveVariants(NameResolution.kt:518)
	at org.rust.lang.core.resolve.NameResolutionKt.processPathResolveVariants(NameResolution.kt:343)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt$resolvePath$result$1.invoke(RsPathReferenceImpl.kt:85)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt$resolvePath$result$1.invoke(RsPathReferenceImpl.kt)
	at org.rust.lang.core.resolve.ProcessorsKt.collectPathResolveVariants(Processors.kt:60)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt.resolvePath(RsPathReferenceImpl.kt:84)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt.resolvePath$default(RsPathReferenceImpl.kt:83)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl$Resolver.invoke(RsPathReferenceImpl.kt:72)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl$Resolver.invoke(RsPathReferenceImpl.kt:70)
	at org.rust.lang.core.resolve.ref.RsResolveCache$resolveWithCaching$$inlined$run$lambda$1.compute(RsResolveCache.kt:94)
	at com.intellij.openapi.util.RecursionManager$1.doPreventingRecursion(RecursionManager.java:113)
	at org.rust.lang.core.resolve.ref.RsResolveCache.resolveWithCaching(RsResolveCache.kt:94)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.advancedCachedMultiResolve(RsPathReferenceImpl.kt:63)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.advancedMultiResolve(RsPathReferenceImpl.kt:59)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.advancedResolve(RsPathReferenceImpl.kt:48)
	at org.rust.lang.core.types.infer.DeclarationsKt.inferTypeReferenceType(Declarations.kt:32)
	at org.rust.lang.core.types.infer.DeclarationsKt.inferTypeReferenceType$default(Declarations.kt:20)
	at org.rust.lang.core.types.ExtensionsKt.getType(Extensions.kt:50)
	at org.rust.lang.core.resolve.RsCachedImplItem$typeAndGenerics$2.invoke(RsCachedImplItem.kt:38)
	at org.rust.lang.core.resolve.RsCachedImplItem$typeAndGenerics$2.invoke(RsCachedImplItem.kt:30)
	at kotlin.SafePublicationLazyImpl.getValue(LazyJVM.kt:107)
	at org.rust.lang.core.resolve.RsCachedImplItem.getTypeAndGenerics(RsCachedImplItem.kt)
	at org.rust.lang.core.resolve.ImplLookup$findSimpleImpls$1.invoke(ImplLookup.kt:372)
	at org.rust.lang.core.resolve.ImplLookup$findSimpleImpls$1.invoke(ImplLookup.kt:178)
	at kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:172)
	at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:132)
	at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:156)
	at kotlin.sequences.TransformingSequence$iterator$1.hasNext(Sequences.kt:176)
	at kotlin.collections.CollectionsKt__MutableCollectionsKt.addAll(MutableCollections.kt:139)
	at org.rust.lang.core.resolve.ImplLookup.rawFindImplsAndTraits(ImplLookup.kt:260)
	at org.rust.lang.core.resolve.ImplLookup.access$rawFindImplsAndTraits(ImplLookup.kt:178)
	at org.rust.lang.core.resolve.ImplLookup$findImplsAndTraits$cached$1.invoke(ImplLookup.kt:239)
	at org.rust.lang.core.resolve.ImplLookup$findImplsAndTraits$cached$1.invoke(ImplLookup.kt:178)
	at org.rust.stdext.Cache$Companion$fromConcurrentMap$1.getOrPut(Cache.kt:41)
	at org.rust.lang.core.resolve.ImplLookup.findImplsAndTraits(ImplLookup.kt:239)
	at org.rust.lang.core.resolve.NameResolutionKt.processAssociatedItems(NameResolution.kt:1145)
	at org.rust.lang.core.resolve.NameResolutionKt.processMethodDeclarationsWithDeref(NameResolution.kt:1125)
	at org.rust.lang.core.resolve.NameResolutionKt.processMethodCallExprResolveVariants(NameResolution.kt:140)
	at org.rust.lang.core.resolve.ref.RsMethodCallReferenceImplKt$resolveMethodCallReferenceWithReceiverType$1.invoke(RsMethodCallReferenceImpl.kt:60)
	at org.rust.lang.core.resolve.ref.RsMethodCallReferenceImplKt$resolveMethodCallReferenceWithReceiverType$1.invoke(RsMethodCallReferenceImpl.kt)
	at org.rust.lang.core.resolve.ProcessorsKt.collectResolveVariantsAsScopeEntries(Processors.kt:94)
	at org.rust.lang.core.resolve.ref.RsMethodCallReferenceImplKt.resolveMethodCallReferenceWithReceiverType(RsMethodCallReferenceImpl.kt:59)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferMethodCallExprType(TypeInferenceWalker.kt:553)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferDotExprType(TypeInferenceWalker.kt:742)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:161)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType$default(TypeInferenceWalker.kt:142)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferBinaryExprType(TypeInferenceWalker.kt:862)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:170)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType$default(TypeInferenceWalker.kt:142)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferChildExprsRecursively(TypeInferenceWalker.kt:1138)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferChildExprsRecursively(TypeInferenceWalker.kt:1139)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferChildExprsRecursively(TypeInferenceWalker.kt:1139)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferMacroExprType0(TypeInferenceWalker.kt:1097)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferMacroExprType(TypeInferenceWalker.kt:1049)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:175)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType$default(TypeInferenceWalker.kt:142)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.processStatement(TypeInferenceWalker.kt:138)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:106)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType$default(TypeInferenceWalker.kt:90)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferBlockExprType(TypeInferenceWalker.kt:57)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:163)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferMatchExprType(TypeInferenceWalker.kt:801)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:168)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferTypeCoercableTo(TypeInferenceWalker.kt:189)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:112)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferTypeCoercableTo(TypeInferenceWalker.kt:88)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferFnBody(TypeInferenceWalker.kt:83)
	at org.rust.lang.core.types.infer.RsInferenceContext.infer(TypeInference.kt:169)
	at org.rust.lang.core.types.infer.TypeInferenceKt$inferTypesIn$1.compute(TypeInference.kt:31)
	at org.rust.lang.core.types.infer.TypeInferenceKt$inferTypesIn$1.compute(TypeInference.kt)
	at com.intellij.openapi.util.RecursionManager$1.doPreventingRecursion(RecursionManager.java:113)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:71)
	at org.rust.openapiext.UtilsKt.recursionGuard(utils.kt:62)
	at org.rust.lang.core.types.infer.TypeInferenceKt.inferTypesIn(TypeInference.kt:31)
	at org.rust.lang.core.types.ExtensionsKt$inference$1.compute(Extensions.kt:66)
	at com.intellij.psi.util.CachedValuesManager.lambda$getCachedValue$0(CachedValuesManager.java:148)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:54)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$1(CachedValueBase.java:240)
	at com.intellij.openapi.util.RecursionManager$1.doPreventingRecursion(RecursionManager.java:113)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:71)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:241)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:43)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:64)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:147)
	at org.rust.lang.core.types.ExtensionsKt.getInference(Extensions.kt:65)
	at org.rust.lang.core.types.ExtensionsKt.getInference(Extensions.kt:140)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.advancedMultiResolve(RsPathReferenceImpl.kt:58)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.multiResolve(RsPathReferenceImpl.kt:51)
	at com.intellij.psi.PsiPolyVariantReferenceBase.resolve(PsiPolyVariantReferenceBase.java:47)
	at org.rust.lang.core.resolve.ref.RsReferenceBase.resolve(RsReferenceBase.kt:28)
	at org.rust.lang.core.types.ExtensionsKt.getDeclaration(Extensions.kt:94)
	at org.rust.ide.hints.RsInlayTypeHintsProvider$getCollectorFor$1.presentTypeForPat(RsInlayTypeHintsProvider.kt:147)
	at org.rust.ide.hints.RsInlayTypeHintsProvider$getCollectorFor$1.presentVariable(RsInlayTypeHintsProvider.kt:123)
	at org.rust.ide.hints.RsInlayTypeHintsProvider$getCollectorFor$1.collect(RsInlayTypeHintsProvider.kt:97)
	at com.intellij.codeInsight.hints.CollectorWithSettings.collectHints(InlayHintsUtils.kt:64)
	at com.intellij.codeInsight.hints.InlayHintsPass$doCollectInformation$1.process(InlayHintsPass.kt:32)
	at com.intellij.codeInsight.hints.InlayHintsPass$doCollectInformation$1.process(InlayHintsPass.kt:15)
	at com.intellij.concurrency.JobLauncherImpl.lambda$null$2(JobLauncherImpl.java:115)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:585)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:531)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:59)
	at com.intellij.concurrency.JobLauncherImpl.lambda$processImmediatelyIfTooFew$3(JobLauncherImpl.java:111)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:791)
	at com.intellij.concurrency.JobLauncherImpl.processImmediatelyIfTooFew(JobLauncherImpl.java:122)
	at com.intellij.concurrency.JobLauncherImpl.invokeConcurrentlyUnderProgress(JobLauncherImpl.java:41)
	at com.intellij.codeInsight.hints.InlayHintsPass.doCollectInformation(InlayHintsPass.kt:23)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.collectInformation(TextEditorHighlightingPass.java:55)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$null$1(PassExecutorService.java:429)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:997)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$2(PassExecutorService.java:422)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:585)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:531)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:59)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.doRun(PassExecutorService.java:421)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$run$0(PassExecutorService.java:397)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:167)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:157)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.run(PassExecutorService.java:395)
	at com.intellij.concurrency.JobLauncherImpl$VoidForkJoinTask$1.exec(JobLauncherImpl.java:161)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```

**Comment by Undin** [comment]
@vlad20012 could you please take a look?
It seems it's because of our caches
We caught similar exceptions in our tests recently

**Comment by ice1000** [comment]
Happens again for me, reported as https://ea.jetbrains.com/browser/ea_reports/5258089

**Comment by Undin** [comment]
Fixed in #4894
